funsor/integrate.py

```diff
@@ -0,0 +1,152 @@
+from __future__ import absolute_import, division, print_function
+
+from collections import OrderedDict
+from six import integer_types
+
+import torch
+
+import funsor.ops as ops
+from funsor.distributions import Gaussian, Delta
+from funsor.interpreter import interpretation, reinterpret
+from funsor.optimizer import Finitary, optimize
+from funsor.terms import Funsor, Number, Reduce, Variable, eager, reflect
+from funsor.torch import Tensor
+
+
+GROUND_TERMS = (Tensor, Gaussian, Delta, Number, Variable)
+
+
+def _make_base_measure(arg, reduced_vars, normalized=True):
+    if not all(isinstance(d.dtype, integer_types) for d in arg.inputs.values()):
+        raise NotImplementedError("TODO implement continuous base measures")
+
+    sizes = OrderedDict(set((var, dtype) for var, dtype in arg.inputs.items()))
+    terms = tuple(
+        Tensor(torch.ones((d.dtype,)) / float(d.dtype), OrderedDict([(var, d)]))
+        if normalized else
+        Tensor(torch.ones((d.dtype,)), OrderedDict([(var, d)]))
+        for var, d in sizes.items() if var in reduced_vars
+    )
+    return Finitary(ops.mul, terms) if len(terms) > 1 else terms[0]
+
+
+def _find_constants(measure, operands):
+    constants, new_operands = [], []
+    for operand in operands:
+        if frozenset(measure.inputs) & frozenset(operand.inputs):
+            new_operands.append(operand)
+        else:
+            constants.append(operand)
+    return tuple(constants), tuple(new_operands)
+
+
+class Integrate(Funsor):
+
+    def __init__(self, measure, integrand):
```

Comment:
I'm pretty sure we'll want to specify `reduced_vars` so we can compute a batch of integrals. I think this will even be needed for iterated expectation with overlapping variables.